x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/PolicyStepsRegistry.java

```diff
@@ -138,30 +150,54 @@ public LifecyclePolicyMetadata read(StreamInput in, String key) {
                 final String existingPhase = (currentSteps == null || currentSteps.size() == 0) ?
                     "_none_" : currentSteps.get(0).getKey().getPhase();
                 // Retrieve the current phase, defaulting to "new" if no phase is set
-                final String currentPhase = imd.value.getSettings().get(LifecycleSettings.LIFECYCLE_PHASE, "new");
+                final String currentPhase = imd.value.getSettings().get(LifecycleSettings.LIFECYCLE_PHASE,
+                    InitializePolicyContextStep.INITIALIZATION_PHASE);
 
                 if (existingPhase.equals(currentPhase) == false) {
                     logger.debug("index [{}] has transitioned phases [{} -> {}], rebuilding step list",
                         index, existingPhase, currentPhase);
-                    // Only rebuild the index's steps if the phase of the existing steps does not match our index's current phase
-                    final Map<Step.StepKey, Step> steps = stepMap.get(policy);
-
+                    // parse existing phase steps from the phase definition in the index settings
+                    String phaseDef = imd.value.getSettings().get(LifecycleSettings.LIFECYCLE_PHASE_DEFINITION,
+                        InitializePolicyContextStep.INITIALIZATION_PHASE);
+                    final Phase phase;
+                    LifecyclePolicy currentPolicy = lifecyclePolicyMap.get(policy).getPolicy();
+                    final LifecyclePolicy policyToExecute;
+                    if (InitializePolicyContextStep.INITIALIZATION_PHASE.equals(phaseDef)
+                            || TerminalPolicyStep.COMPLETED_PHASE.equals(phaseDef)) {
+                        // It is ok to re-use potentially modified policy here since we are in an initialization or completed phase
+                        policyToExecute = currentPolicy;
+                    } else {
+                        // if the current phase definition describes an internal step/phase, do not parse
+                        try (XContentParser parser = JsonXContent.jsonXContent.createParser(xContentRegistry,
+                            DeprecationHandler.THROW_UNSUPPORTED_OPERATION, phaseDef)) {
+                            phase = Phase.parse(parser, currentPhase);
+                        } catch (IOException e) {
+                            logger.error("failed to configure phase [" + currentPhase + "] for index [" + index.getName() + "]", e);
+                            indexPhaseSteps.remove(index);
+                            continue;
+                        }
+                        Map<String, Phase> phaseMap = new HashMap<>(currentPolicy.getPhases());
+                        if (phase != null) {
+                            phaseMap.put(currentPhase, phase);
+                        }
+                        policyToExecute = new LifecyclePolicy(currentPolicy.getType(), currentPolicy.getName(), phaseMap);
```

Comment:
Why do we need to create a fake policy here with only 1 phase in it?

Reply:
@talevy  and I spoke about this and although this will want to change at some point we can't change to compile directly from the phase right now whilst the PhaseAfterStep still exists (which is being removed in a different task).